### ISSUE TITLE
Create native function to compute hashcodes for value types

### ISSUE THREAD

**Issue by ehrenjulzert** [eclipse-openj9/openj9#15768]
Currently the only way to compute a hashcode for value types is by calling the `primitiveObjectHashCode` method of the `PrimitiveObjectMethods` class. This is an issue because `primitiveObjectHashCode` cannot be called in non-Java threads (and on top of that it is slow as the hashcode is not cached).

For https://github.com/eclipse-openj9/openj9/issues/15529

**Comment by hangshao0** [comment]
This issue is for project Valhalla only. I see it is tracked under project Loom and Java 19. 

I've removed it from Project Loom. 

@tajila I see you added it to Java 19 milestone, which should be removed ?

**Comment by ehrenjulzert** [comment]
To implement this I'll be using the Murmur3 algorithm that we already use to calculate identity hashcodes

**Comment by tajila** [comment]
Ya, essentially you need to iterate the fields

**Comment by theresa-m** [comment]
This should fix ValueTypeTests:testValueClassHashCode and some tests in test/jdk/valhalla/valuetypes/ValueObjectMethodsTest.java https://github.com/eclipse-openj9/openj9/issues/23172.

**Comment by theresa-m** [comment]
@aditis11 please take a look when you have the cycles.

**Comment by AditiS11** [comment]
Will work on it.

**Comment by theresa-m** [comment]
From the previous discussion https://github.com/eclipse-openj9/openj9/pull/15880 I believe there are two steps involved:
1. Update VM_ValueTypeHelpers::isSubstitutable and convertValueObjectAtOffsetToHash to use iterative approaches so there is no risk of a stack overflow. @AditiS11 confirmed that ri is still throwing a stackoverflow but this is not a good final solution for OpenJ9 https://github.com/eclipse-openj9/openj9/pull/15880#issuecomment-1248374930.
2. Add support for all other methods of gc object tracing in addition to forwarded pointers. We'll need some help from the gc team to understand this further.

fyi @AditiS11 @hangshao0 @dmitripivkine

**Comment by hangshao0** [comment]
> Update VM_ValueTypeHelpers::isSubstitutable and convertValueObjectAtOffsetToHash to use iterative approaches 

As both functions need to traverse all the fields, it worth consider adding a helper function that can be used by isSubstitutable and calculating hash.

**Comment by dmitripivkine** [comment]
The problem with previous approaches calculation of the hash code required access to different objects and it can be challenging. Currently GC requests hash code calculation during move an object with "hashed" bit set. So, at new location copy of the object should be extended by hash slot and it should be inflated. 
If we take Scavenger as example to demonstrate complexity there is a fundamental issue to access to another object in the Nursery. Assume Scavenger needs to copy object A happen to be "hashed" and VT. But if it is refers to another object B (VT as well) located in Nursery and has Forwarded Header installed already we don't know was object B copy operation complete at this point. So, we can not (and should not) use Evacuate copy due installing of FP destroys object first slot, but also can not use it in Survivor because copy operation (by another thread) might not be complete. For STW Scavenger we can postpone hash slot calculation in theory - add hash slot only inflate later after copy complete, but it requires to remember such objects. If GC builds a list of them it requires to add hidden "next" slot, which is equivalent for object size to add extended hash code for each VT object at the first place.

**Comment by dmitripivkine** [comment]
The problem can be resolved if hash code can be pre-calculated at the time of hashing and GC just request it when need to inflate  hash slot. It is challenging as well.

**Comment by theresa-m** [comment]
If my understanding is correct the hashed bit is not set for an object unless the Object.hashCode is called by a user. In that case I think the first solution makes more sense since there would at least be a chance that allocating a list of objects for STW would use less memory then precalculating all the hash codes.

Currently all of the migrated value classes in the JDK all have primitive instance fields so these would not have to be stored in the list since their hash code does not rely on other objects. These classes will likely be the most common uses (@hangshao0 can confirm).

**Comment by hangshao0** [comment]
> Currently all of the migrated value classes in the JDK all have primitive instance fields

Yes, except for Optional. However, Optional overrides hashCode() method so we don't have to worry about it.

All the fields in a value type object is final. If the filed is primitive or value type (non-reference type), its value will be the same before and after GC. If the field is a reference type, its hashcode will be the same before and after GC. We might consider calculating the hashcode based on the value of non-reference fields and identity hashcode of reference fields.

**Comment by dmitripivkine** [comment]
>In that case I think the first solution makes more sense since...

Which solution are you talking about? An original one [we have implemented](https://github.com/eclipse-openj9/openj9/pull/15880/changes/ffa38ab4475b8571d9c7e54a30c5dbe35c148a43)?

**Comment by AditiS11** [comment]
Re: Using identity hashcode for reference fields
If we use identity hashcode for reference fields, two value type objects with the same primitive field values and reference fields containing equal content will have different hashcodes as the reference objects will have different identity hashcodes. 
Is this acceptable?

**Comment by theresa-m** [comment]
> Which solution are you talking about? An original one [we have implemented](https://github.com/eclipse-openj9/openj9/pull/15880/changes/ffa38ab4475b8571d9c7e54a30c5dbe35c148a43)?

I mean doing 
> For STW Scavenger we can postpone hash slot calculation in theory - add hash slot only inflate later after copy complete, but it requires to remember such objects. If GC builds a list of them it requires to add hidden "next" slot, which is equivalent for object size to add extended hash code for each VT object at the first place.

I was under the impression the implementation in https://github.com/eclipse-openj9/openj9/pull/15880/changes/ffa38ab4475b8571d9c7e54a30c5dbe35c148a43 does not work in all cases.

**Comment by theresa-m** [comment]
> Re: Using identity hashcode for reference fields If we use identity hashcode for reference fields, two value type objects with the same primitive field values and reference fields containing equal content will have different hashcodes as the reference objects will have different identity hashcodes. Is this acceptable?

Yes. This follows the existing behaviour of the ri.

**Comment by dmitripivkine** [comment]
After offline discussion with @hangshao0 we concluded with this flow of operations:

- Hashing VT object sets “has been hashed” bit.
- Flow in GC does not change (expanding hash slot during first copy and set “has been moved”) except hash calculating function returns 0 for VT objects.
- Later on when hash code is requested, VM discovers 0 in the hash slot, recalculate value and update the slot.
- Actual VT object hash calculation should not be called during copy collectors (needs split of API for functions called from GC and VM)
- VM can request preservation of hash codes for recursively referenced VT objects as optimization based on the same rules (setting “has been hashed” bit to them) if like, not mandatory.
- we should not forget to check that actual hash code it not equal 0 (or just recalculate/update it every time)

Also VM code might need better support from GC like external function to calculate hash. Current prototype creates local copy of the barrier. This is not acceptable. GC should be free to use any barrier implementation it needs. May be it is convenient to have GC helper to calculate VT object hash code and preserve it if hash slot is inflated.

**Comment by keithc-ca** [comment]
The hashcodes for instances of value types should not depend on identity, so GC should not need to be involved (any more than it would be for a field of type `int` or `long`).

**Comment by dmitripivkine** [comment]
GC involvement is required if VM wants to store calculated hash code in inflated object hash slot.

**Comment by dmitripivkine** [comment]
If VM does not want to store hash in the object it should not set "has been hashed" bit, so GC would not be aware about VT hashing at all. In this case VM just need to respect VM Access during VT object hash calculation and use existing macro (with implemented barrier calls inside). GC hash code calculation path should not be changed.

**Comment by keithc-ca** [comment]
It's my understanding that the primary motivation for storing a hash code in inflated object hash slot is so that value can be stable even though an object (or others to which it refers) may move during a GC cycle. That is not an issue for value types which should not contain object references and so are unaffected by the movement of any object (including themselves).

I expect the vast majority of value types to be relatively small and so the cost of recomputing a hash code should be less concerning.

**Comment by hangshao0** [comment]
Not sure how big user defined value types will be and how frequently hashCode() will be called on them. I still prefer caching the hashcode in the object to avoid calculating the hash repeatedly. A threshold on the value type size could be introduced to enable caching of the hashcode.

**Comment by theresa-m** [comment]
@keithc-ca I tried this out last week to double check for myself and value type hashes can rely on the identity of thier identity class fields.

```
class MyTest {
        public static void main(String[] args) {
                R r1 = new R(0);
                R r2 = new R(0);
                System.out.println("Identity class: " + (r1.hashCode() == r2.hashCode()));

                V v1 = new V(0);
                V v2 = new V(0);
                System.out.println("Value class: " + (v1.hashCode() == v2.hashCode()));

                VwithV vwithv1 = new VwithV(0);
                VwithV vwithv2 = new VwithV(0);
                System.out.println("Value class with value class field: " + (vwithv1.hashCode() == vwithv2.hashCode()));

                VwithR vwithr1 = new VwithR(0);
                VwithR vwithr2 = new VwithR(0);
                System.out.println("Value class with identity class field: " + (vwithr1.hashCode() == vwithr2.hashCode()));
        }
}

class R {
        int i;
        R(int i) { this.i = i; }
}

value class V {
        int i;
        V(int i) {this.i = i;}
}

value class VwithV {
        int i;
        V v;
        VwithV(int i) {this.i = i; this.v = new V(i);}
}

value class VwithR {
        int i;
        R r;
        VwithR(int i) {this.i = i; this.r = new R(i);}
}
```

Results using the ri: 
```
Identity class: false
Value class: true
Value class with value class field: true
Value class with identity class field: false
```

**Comment by keithc-ca** [comment]
That's disappointing. I'm a little surprised that `VwithR` is even legal. While it remains legal, such value classes must use the "object hash slot" to handle situations in which the object to which `VwithR.r` refer is relocated (unless there's no guarantee that hash codes for such types are stable). What does the RI say about that? Is the hash code stable, or does it change if `R` objects move?

**Comment by theresa-m** [comment]
System.identityHashCode is used in the calculation which returns a consistent hash code when called multiple times in an application.

**Comment by keithc-ca** [comment]
Yes, but the answer is consistent because the first time it is computed it is cached (at least in OpenJ9) so it isn't affected by references to moving objects.

It's my understanding that the formula for `System.identityHashCode()` is not prescribed by any specification.

I'd still like to know whether `VwithR.hashCode()` returns the same value (in the RI) even if `R` objects move.

**Comment by hangshao0** [comment]
I will be surprised if an Object (either value type or reference type) could have inconsistent hash code value between difference executions of `hashCode()` method.

**Comment by theresa-m** [comment]
I see it in https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/Object.html#hashCode()
> Whenever it is invoked on the same object more than once during an execution of a Java application, the hashCode method must consistently return the same integer, provided no information used in equals comparisons on the object is modified.

System. identityHashCode refers to this definition as well.

**Comment by keithc-ca** [comment]
That's a description of the expectations for user-written `hashCode()` methods. It's fair to assume that `System.identityHashCode()` satisfies those expectations, at least for non-value types. The question here is whether _value_ types should be expected to have `hashCode()` methods that behave that way as well. The test code above doesn't answer that question.

**Comment by tajila** [comment]
> That's a description of the expectations for user-written hashCode() methods. 

This is not true. The `Object::hashcode` contract applies whether hashcode is user defined or not, to all types that subtype from java.lang.Object which includes value types.

> The question here is whether value types should be expected to have hashCode() methods that behave that way as well.

Yes, they must.

Some more details from https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/lang/Object.html#hashCode()

```
If two objects are equal according to the equals method, then calling the hashCode method on each of the two objects must produce the same integer result.
```

This is necessary for collections APIs to function properly. So all types in the Java type system that are objects must adhere to this contract.

**Comment by keithc-ca** [comment]
https://openjdk.org/jeps/401 suggests that the behavior of `VwithR` may change:

> The inherited implementation of Object.hashCode computes a hash from the object's field values. (This value can also be computed via System.identityHashCode.) As usual, the hashCode method should be overridden by a value class whenever it overrides equals.

> If the equals and hashCode methods have not already been overridden, they will behave differently after migration. A good migration candidate will want to override these methods beforehand so that their behavior does not depend on identity.

**Comment by tajila** [comment]
No it doesn't. It states that if the default behaviour of `equals` (which for value types performs a comparison based on objects state) is not desired then the value class should override equals.

> The inherited implementation of Object.equals uses == to compare objects. For value objects, this tests for statewise equivalence. This might be the right equals behaviour for a value class, but if it isn't then the class author should override equals.

It then says, `hashcode` for value types is computed by the object state (similar to equals) and it should be overridden if equals is overridden. This confirms what we discussed earlier.

In the second section you posted it talks about existing types being migrated to value types and warns that if `equals` and `hashcode` is not overridden the behaviour may change. 

It doesn't suggest that the `VwithR` example will behave differently, it actually confirms that the behaviour observed is intended.

It suggests that if `VwithR` was migrated to a value type the writer of the class may want to override `equals` and `hashcode` or they would see the behaviour described in https://github.com/eclipse-openj9/openj9/issues/15768#issuecomment-3945121431.

**Comment by keithc-ca** [comment]
I expect `javac` to generate `equals()` and `hashCode()` for `value` types, like it does for `record` types. The compiler currently does not do that, so we may not have a clear picture of the final behavior. If it does generate `hashCode()` following the rules described in JEP401, the behavior of `VwithR` will cease to depend on the location of the referenced `R` object.

**Comment by tajila** [comment]
> I expect javac to generate equals() and hashCode() for value types, like it does for record types

Its not correct to assume javac will create all value types. But even if this is true, then the behaviour will depend on the implementation of those methods. 

This doesn't change what we are discussing here because we are discussing the behaviour when `hashCode` and `equals` is not overridden and instead based on j.l.Object behaviour. 

> If it does generate hashCode() following the rules described in JEP401, the behavior of VwithR will cease to depend on the location of the referenced R object.

Nothing in JEP401 suggests this.

**Comment by keithc-ca** [comment]
The default (inherited) behavior of `hashCode()` for objects with identity is the same as `System.identityHashCode()`. That makes no sense for value types which have no identity.

In JEP401 is says:

> The inherited implementation of Object.hashCode computes a hash from the object's field values.

I read that to mean it would involve the field _hash_ values, like the sample `LazySubstring` shown there.

**Comment by tajila** [comment]
> The default (inherited) behavior of hashCode() for objects with identity is the same as System.identityHashCode(). That makes no sense for value types which have no identity.

Whether it makes sense is irrelevant. The relevant question is "what is the behaviour of `Object::hashcode` (and by extension `System.identityHashCode()`) when applied to valuetypes?". 

As stated in the JEP, methods that are traditionally identity sensitive will rely on the objects fields instead.

The contract for hashcode remains the same, whether it is a valuetype or not.

> Whenever it is invoked on the same object more than once during an execution of a Java application, the hashCode method must consistently return the same integer..

> If two objects are equal according to the [equals](https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/lang/Object.html#equals(java.lang.Object)) method, then calling the hashCode method on each of the two objects must produce the same integer result.

Given that the location of referenced objects can move, this cannot be relied on to compute the hashcode.

>>   The inherited implementation of Object.hashCode computes a hash from the object's field values.

> I read that to mean it would involve the field hash values, like the sample LazySubstring shown there.

LazySubstring does not use the default hashcode implementation so it is not relevant to this discussion.

**Comment by hangshao0** [comment]
For the hash code calculation described in https://github.com/eclipse-openj9/openj9/issues/15768#issuecomment-3936543602

> Flow in GC does not change (expanding hash slot during first copy and set “has been moved”) except hash calculating function returns 0 for VT objects.
> Later on when hash code is requested, VM discovers 0 in the hash slot, recalculate value and update the slot.

Like what we do to identity object, GC could inflated the object with an extra hashlot to cache the hashcode. The difference is that for value type objects, GC will put a temporary 0 and the next hash code request will update it to real hashcode.

Not sure if the temporary 0 in the hashslot (stored at the backfillOffset from the object) will be a problem for JIT. Does JIT read this slot directly assuming it is the hashcode ? @hzongaro @a7ehuo 
https://github.com/eclipse-openj9/openj9/blob/8defdc3c429a09b12614b3337e3a29b41f89c233/runtime/oti/ObjectHash.hpp#L309

We should be fine if the answer is no.

**Comment by theresa-m** [comment]
Thanks @AditiS11 for your work on this!
